aws-xray-recorder-sdk-core/src/test/java/com/amazonaws/xray/contexts/CustomSegmentContextTest.java

```diff
@@ -124,7 +124,7 @@ void testGlobalMapSegmentContext() {
             list.add(i);
         }
 
-        list.parallelStream().forEach(e -> {
+        list.forEach(e -> {
             AWSXRay.setTraceEntity(test);
             AWSXRay.createSubsegment("parallelPrint", (subsegment) -> {
```

Comment:
`print`? Though if the test is trying to test parallel processing, maybe this isn't a good fix to remove the parallelism. 

Could you post the stack trace of the test failure? I don't quite where the error is - parallel threads will each have their threadlocalcontext, and we set `.setTraceEntity` at the beginning of each runnable (technically we're setting `traceEntity` many times more than we need to but don't think it should hurt). I'm not finding where the context could be replaced with null or anything particularly wrong with the test code right now.

Reply:
Here's the stacktrace:

```
[junit]      Testcase: testGlobalMapSegmentContext(com.amazonaws.xray.contexts.CustomSegmentContextTest):    Caused an ERROR
[junit]      null
[junit]      java.lang.NullPointerException
[junit]          at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[junit]          at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598)
[junit]          at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
[junit]          at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
[junit]          at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
[junit]          at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
[junit]          at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
[junit]          at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
[junit]          at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:650)
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest.testGlobalMapSegmentContext(CustomSegmentContextTest.java:109)
[junit]      Caused by: java.lang.NullPointerException
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest$GlobalMapSegmentContext.beginSubsegment(CustomSegmentContextTest.java:33)
[junit]          at com.amazonaws.xray.AWSXRayRecorder.beginSubsegment(AWSXRayRecorder.java:504)
[junit]          at com.amazonaws.xray.AWSXRayRecorder.createSubsegment(AWSXRayRecorder.java:303)
[junit]          at com.amazonaws.xray.AWSXRay.createSubsegment(AWSXRay.java:64)
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest.lambda$testGlobalMapSegmentContext$1(CustomSegmentContextTest.java:111)
[junit]          at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[junit]          at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
[junit]          at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
[junit]          at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
[junit]          at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
[junit]          at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[junit]          at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[junit]          at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[junit]          at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

Reply:
This test doesn't use `ThreadLocalContext`, it's testing a custom segment context to replace thread local. However the custom context is just a map of thread IDs to X-Ray entities so it *should* be functionally the same as ThreadLocal. Since the test was testing a replaced context, I didn't think removing the parallelism would be a big deal but I think you're right that we should keep it.

I'm also having a tough time finding when the context would be set to null, but it must be happening somewhere since customers have reported this test sporadically failing with the above stacktrace. Will continue investigating.

Reply:
Customers also reported a `SubsegmentNotFoundException`, which has the same root cause as the NPE (the map has a null entry), but it's just caught when ending the subsegment:

```
Testcase: testGlobalMapSegmentContext(com.amazonaws.xray.contexts.CustomSegmentContextTest):    Caused an ERROR
[junit]      com.amazonaws.xray.exceptions.SubsegmentNotFoundException: Failed to end subsegment: subsegment cannot be found.
[junit]      com.amazonaws.xray.exceptions.SubsegmentNotFoundException: com.amazonaws.xray.exceptions.SubsegmentNotFoundException: Failed to end subsegment: subsegment cannot be found.
[junit]          at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[junit]          at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:593)
[junit]          at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
[junit]          at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
[junit]          at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)
[junit]          at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
[junit]          at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
[junit]          at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
[junit]          at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:650)
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest.testGlobalMapSegmentContext(CustomSegmentContextTest.java:109)
[junit]      Caused by: com.amazonaws.xray.exceptions.SubsegmentNotFoundException: Failed to end subsegment: subsegment cannot be found.
[junit]          at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
[junit]          at com.amazonaws.xray.strategy.RuntimeErrorContextMissingStrategy.contextMissing(RuntimeErrorContextMissingStrategy.java:17)
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest$GlobalMapSegmentContext.endSubsegment(CustomSegmentContextTest.java:55)
[junit]          at com.amazonaws.xray.AWSXRayRecorder.endSubsegment(AWSXRayRecorder.java:520)
[junit]          at com.amazonaws.xray.AWSXRayRecorder.createSubsegment(AWSXRayRecorder.java:312)
[junit]          at com.amazonaws.xray.AWSXRay.createSubsegment(AWSXRay.java:64)
[junit]          at com.amazonaws.xray.contexts.CustomSegmentContextTest.lambda$testGlobalMapSegmentContext$1(CustomSegmentContextTest.java:111)
[junit]          at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
[junit]          at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
[junit]          at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
[junit]          at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)
[junit]          at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)
[junit]          at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[junit]          at java.util.concurrent.ForkJoinPool$WorkQueue.execLocalTasks(ForkJoinPool.java:1040)
[junit]          at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1058)
[junit]          at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[junit]          at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

Reply:
Really unsure how to find the root cause of this problem or fix it without removing parallelism (which would make the test less useful). I tried increasing the list size to 10000 and it still ran fine on my machine. I guess it comes down to different JVMs creating threads differently.

Reply:
Oh I didn't notice this is testing a custom segment, now I looked at the whole file... Yeah I think it's not safe to write to a `Map` or any resizable data structure really - if a thread resizes, the other threads may not see the change and things get bad. Changing to `ConcurrentHashMap` would probably fix it, but removing the parallelness seems ok to me too, I realized it's not a test of parallelism. Let's change the `parallelPrint` string too if we go that route.

Reply:
Yeah seeing as I stared at this for over an hour and couldn't come up with a sane way to fix while maintaining parallelism I'll just stick with removing `parallelStream` and renaming the string, since removing whatever flakiness there was in this test is most important.